Add docs for how airflow manages packages and imports (#10303)
mistercrunch/Airflow: docs/packages.rst

 .. Licensed to the Apache Software Foundation (ASF) under one
    or more contributor license agreements.  See the NOTICE file
    distributed with this work for additional information
    regarding copyright ownership.  The ASF licenses this file
    to you under the Apache License, Version 2.0 (the
    "License"); you may not use this file except in compliance
    with the License.  You may obtain a copy of the License at

 ..   http://www.apache.org/licenses/LICENSE-2.0

 .. Unless required by applicable law or agreed to in writing,
    software distributed under the License is distributed on an
    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, either express or implied.  See the License for the
    specific language governing permissions and limitations
    under the License.

> Loading Packages in Python and Airflow

Reviewer: ```suggestion
Loading modules
```
or 
```suggestion
Module management
```
The title should consist of 2-3 words to fit correctly in the menu.

======================================

> 

Reviewer: ```suggestion
Airflow allows you to use your own Python modules in the DAG and in the Airflow configuration. The following article will describe how you can create your own module so that Airflow can load it correctly, as well as diagnose problems when modules are not loaded properly.

This article is the last one for you if you need to adapt Airflow to the needs of your organization.
```

> How does Python load packages?

Reviewer: If possible, we should avoid questions in headlines.

------------------------------

> Python's `sys <https://docs.python.org/3/library/sys.html>`_ package provides key variables and functions that are used

Reviewer: ```
Airflow uses the standard Python loader to load additional modules. All modules that can find and load it can also be used.
```

> Python's `sys <https://docs.python.org/3/library/sys.html>`_ package provides key variables and functions that are used
> or maintained by the interpreter. One such variable is ``sys.path``. It's a list of directories which is searched by
> Python during imports. for example,

Reviewer: ```suggestion
The list of directories from which Python tries to load the module is given by the variable: any:`sys.path`. Python really tries to `intelligently determine the contents of <https://stackoverflow.com/a/38403654>` __ of this variable, including depending on the operating system and how Python is installed.

You can check the contents of this variable for the current Python environment by running an interactive terminal as in the example below:
```

.. code-block:: pycon

    >>> import sys
    >>> from pprint import pprint
    >>> pprint(sys.path)
    ['',
     '/home/arch/.pyenv/versions/3.7.4/lib/python37.zip',
     '/home/arch/.pyenv/versions/3.7.4/lib/python3.7',
     '/home/arch/.pyenv/versions/3.7.4/lib/python3.7/lib-dynload',
     '/home/arch/venvs/airflow/lib/python3.7/site-packages']

``sys.path`` is initialized during program startup. The first precedence is given to the current directory,
i.e, ``path[0]`` is the directory containing the current script that was used to invoke or an empty string in case
it was an interactive shell. Second precedence is given to the ``PYTHONPATH``, followed by installation-dependent
default paths which is managed by `site <https://docs.python.org/3/library/site.html#module-site>`_ module.

> 

Reviewer: > In the variable: `sys.path` there is a directory` site-packages`` which contains **external packages * installed, which means you can install packages with `pip` or` anaconda` `and you can use them in Airflow. In the __TODO__ section, you will learn how to create your own simple installable package.

> You can specify additional directories to be added to ``sys.path`` with the ** environment variable. :envvar:`PYTHONPATH` **. See section TODO for more information.

``sys.path`` can also be modified during a Python session, by simply using append (for example, ``sys.path.append("/path/to/custom/package")``). Python will start searching in the newer paths once they're added. Airflow makes use
of this feature as described in the next section.

How does Airflow modify this behavior?
--------------------------------------
Airflow adds three additional directories to ``sys.path``:

- ``conf.get('core', 'dags_folder')``

> - ``conf.get('core', 'airflow_home')``

Reviewer: ```suggestion
- ``conf.get('core', 'airflow_home')/config``
```

- ``conf.get('core', 'plugins_folder')``

> 

Reviewer: ```suggestion
You can use them similar to directories specified with the environment variable :envvar:`PYTHONPATH`.
```

When and how you can affect the module loading mechanism?
---------------------------------------------------------

How to create python package with operators/plugins?
----------------------------------------------------

> 

Reviewer: Install the required packages:

Setuptools: Setuptools is a package development process library designed for creating and distributing Python packages.
Wheel: The Wheel package provides a bdist_wheel command for setuptools. It creates .whl file which is directly installable through the pip install command. We'll then upload the same file to pypi.org.

1. pip install --upgrade pip setuptools wheel

2. Create a directory for all files. In our case, we will call it `airflow_operators`
3. Create setup.py

```python
import setuptools

setuptools.setup(
	name='airflow_operators',  
)
```
3. Build a wheel
```
python setup.py bdist_wheel
```

4. You can install the .whl file using pip:
```
pip install dist/AAAAA
```

See: https://packaging.python.org/tutorials/packaging-projects/

How to use PYTHONPATH?
----------------------

How do you check the contents of the sys.path variable?
-------------------------------------------------------

> 

Reviewer: To check the current contents of the variable: any: `sys.path` in your environment, you can run the use ``airflow info``. An example of the contents of the sys.path variable specified by this command may be as follows:
```
Python PATH: [/usr/local/bin:/opt/airflow:/usr/local/lib/python36.zip:/usr/local/lib/python3.6:/usr/local/lib/python3.6/lib-dynload:/usr/local/lib/python3.6/site-packages:/opt/airflow/airflow/providers/google/cloud/example_dags:/root/airflow/config:/root/airflow/plugins]
```
